src/liballoc_frame/lib.rs

```diff
@@ -47,26 +48,29 @@ const MIN_ALIGN: usize = 16;
 const CHUNK_SIZE: usize = 4096 * 16;
 const CHUNK_ALIGN: usize = 4096;
 
-static mut HEAP: *mut u8 = ptr::null_mut();
-static mut HEAP_LEFT: usize = 0;
+static HEAP: AtomicPtr<u8> = AtomicPtr::new(ptr::null_mut());
+static HEAP_LEFT: AtomicUsize = AtomicUsize::new(0);
 
 #[no_mangle]
 pub extern "C" fn __rust_allocate(size: usize, align: usize) -> *mut u8 {
     let new_align = if align < MIN_ALIGN { MIN_ALIGN } else { align };
     let new_size = (size + new_align - 1) & !(new_align - 1);
 
     unsafe {
-        if new_size < HEAP_LEFT {
-            HEAP_LEFT -= new_size;
-            let p = HEAP;
-            HEAP = HEAP.offset(new_size as isize);
-            return p;
-        } else if new_size > CHUNK_SIZE {
+        if new_size > CHUNK_SIZE {
             return imp::allocate(size, align);
+        }
+        
+        let heap = HEAP.load(Ordering::SeqCst);
+        let heap_left = HEAP_LEFT.load(Ordering::SeqCst);
```

Comment:
Makes sense. I have little experience with atomics, as you may have guessed. I would rather use a `Mutex` or `thread_local!`, but neither of those are in core. I'll make it use a spinlock which should hopefully be correct.